sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/ShouldImplementExportedInterfacesBase.cs

```diff
@@ -0,0 +1,126 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.Common
+{
+    public abstract class ShouldImplementExportedInterfacesBase : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4159";
+        protected const string MessageFormat = "{0} '{1}' on '{2}' or remove this export attribute.";
+        protected const string ActionForInterface = "Implement";
+        protected const string ActionForClass = "Derive from";
+
+        internal static readonly ISet<KnownType> ExportAttributes =
+            new HashSet<KnownType>
+            {
+                KnownType.System_ComponentModel_Composition_ExportAttribute,
+                KnownType.System_ComponentModel_Composition_InheritedExportAttribute,
+            };
+    }
+
+    public abstract class ShouldImplementExportedInterfacesBase<TArgumentSyntax, TExpressionSyntax, TClassSyntax>
+        : ShouldImplementExportedInterfacesBase
+        where TArgumentSyntax : SyntaxNode
+        where TExpressionSyntax : SyntaxNode
+        where TClassSyntax : SyntaxNode
+    {
+        protected static bool IsOfExportType(ITypeSymbol type, INamedTypeSymbol exportedType) =>
+            type.GetSelfAndBaseTypes()
+                .Union(type.AllInterfaces)
+                .Any(currentType => currentType.Equals(exportedType));
+
+        protected INamedTypeSymbol GetExportedTypeSymbol(SeparatedSyntaxList<TArgumentSyntax>? attributeArguments,
+            SemanticModel semanticModel)
+        {
+            if (attributeArguments == null)
+            {
+                return null;
+            }
+
+            var arguments = attributeArguments.Value;
+            if (arguments.Count == 0 ||
+                arguments.Count > 2)
+            {
+                return null;
+            }
+
+            var argumentSyntax = GetArgumentFromNamedArgument(arguments) ??
+                GetArgumentFromSingleArgumentAttribute(arguments) ??
+                GetArgumentFromDoubleArgumentAttribute(arguments, semanticModel);
+            var typeOfExpression = GetExpression(argumentSyntax);
+
+            var exportedTypeSyntax = GetExportedTypeSyntax(typeOfExpression);
+            if (exportedTypeSyntax == null)
+            {
+                return null;
+            }
+
+            return semanticModel.GetSymbolInfo(exportedTypeSyntax).Symbol as INamedTypeSymbol;
+        }
+
+        private TArgumentSyntax GetArgumentFromNamedArgument(IEnumerable<TArgumentSyntax> arguments) =>
+            arguments.FirstOrDefault(x => GetIdentifier(x) == "contractType");
+
+        private TArgumentSyntax GetArgumentFromSingleArgumentAttribute(IEnumerable<TArgumentSyntax> arguments)
+        {
+            if (arguments.ElementAtOrDefault(1) != null)
```

Comment:
Wouldn't it be clearer to use `arguments.Count()`?

Reply:
I assume this also checks that the item in position 1 is not null itself.